pkg/analyze/archivereader_test.go

```diff
@@ -0,0 +1,373 @@
+  conditions:
+  - type: Ready
+    status: "True"
+  containerStatuses:
+  - name: webhook-server
+    ready: true
+    state:
+      running: {}
+`, "\n"))},
+				},
+			},
+			expectedError: nil,
+			expectedIndexerObj: func() cache.Indexer {
+				indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+
+				pod := &corev1.Pod{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Pod",
+						APIVersion: "v1",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "webhook-server-7cbd674df9-782zm",
+						Namespace: "scylla-operator",
+						Labels: map[string]string{
+							"app.kubernetes.io/name": "webhook-server",
+						},
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "webhook-server",
+								Image: "docker.io/scylladb/scylla-operator:latest",
+								Ports: []corev1.ContainerPort{
+									{
+										ContainerPort: 5000,
+										Protocol:      corev1.ProtocolTCP,
+									},
+								},
+							},
+						},
+						Volumes: []corev1.Volume{
+							{
+								Name: "cert",
+								VolumeSource: corev1.VolumeSource{
+									Secret: &corev1.SecretVolumeSource{
+										SecretName: "scylla-operator-serving-cert",
+									},
+								},
+							},
+						},
+					},
+					Status: corev1.PodStatus{
+						Conditions: []corev1.PodCondition{
+							{
+								Type:   corev1.PodConditionType("Ready"),
+								Status: corev1.ConditionTrue,
+							},
+						},
+						ContainerStatuses: []corev1.ContainerStatus{
+							{
+								Name:  "webhook-server",
+								Ready: true,
+								State: corev1.ContainerState{
+									Running: &corev1.ContainerStateRunning{},
+								},
+							},
+						},
+					},
+				}
+
+				if err := indexer.Add(pod); err != nil {
+					panic(err)
+				}
+
+				return indexer
+			}(),
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			indexers, err := IndexersFromArchive(tc.archive)
+
+			if tc.testType == "error check" {
+				if !errorsEqual(err, tc.expectedError) {
+					klog.Errorf("got error %v expected error %v", err, tc.expectedError)
+					t.Fatal(err)
+				}
+				count := 0
+				for _, indexer := range indexers {
+					count += len(indexer.List())
+				}
+				if count != tc.ExpectedIndexerObjectsNum {
+					t.Errorf("expected and got indexer items differ (got %d expected %d)", count, tc.ExpectedIndexerObjectsNum)
+				}
+			}
+			if tc.testType == "data check" {
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				//fmt.Println(reflect.TypeOf(indexers[reflect.TypeOf(&corev1.Pod{})]))
+				actualJson, err := json.MarshalIndent(indexers[reflect.TypeOf(&corev1.Pod{})].List(), "", " ")
+				if err != nil {
+					t.Errorf("marshall1 %s", err)
+					panic("marshall failed")
+				}
+				expectedJson, err := json.MarshalIndent(tc.expectedIndexerObj.List(), "", " ")
+				if err != nil {
+					t.Errorf("marshall2 %s", err)
+
+					panic("marshall failed")
+				}
```

Comment:
1. what's with the panics?
2. this is going to need a better way of comparing the expected'/actual indexers than just comparing an arbitrary type

Reply:
To get you started, below is what I would begin with. You can think about how to cover the indexers better than just comparing the list of items.

```
package analyze

import (
	"reflect"
	"strings"
	"testing"
	"testing/fstest"

	"github.com/google/go-cmp/cmp"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/serializer"
)

func TestIndexersFromFS(t *testing.T) {
	t.Parallel()

	testScheme := runtime.NewScheme()

	err := corev1.AddToScheme(testScheme)
	if err != nil {
		t.Fatal(err)
	}

	testDecoder := serializer.NewCodecFactory(testScheme).UniversalDecoder()

	tt := []struct {
		name                   string
		fs                     fstest.MapFS
		expectedIndexerObjects map[reflect.Type][]interface{}
		expectedError          error
	}{
		{
			name:                   "empty fs is valid",
			fs:                     fstest.MapFS{},
			expectedIndexerObjects: map[reflect.Type][]interface{}{},
			expectedError:          nil,
		},
		{
			name: "single type",
			fs: fstest.MapFS{
				"my-namespace.yaml": &fstest.MapFile{
					Data: []byte(strings.TrimSpace(`
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    annotation-key: annotation-value
  name: my-namespace
spec: {}
status: {}
`)),
				},
			},
			expectedIndexerObjects: map[reflect.Type][]interface{}{
				reflect.TypeOf(&corev1.Namespace{}): {
					&corev1.Namespace{
						TypeMeta: metav1.TypeMeta{
							APIVersion: "v1",
							Kind:       "Namespace",
						},
						ObjectMeta: metav1.ObjectMeta{
							Name: "my-namespace",
							Annotations: map[string]string{
								"annotation-key": "annotation-value",
							},
						},
					},
				},
			},
			expectedError: nil,
		},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			indexers, err := IndexersFromFS(tc.fs, testDecoder)
			if !reflect.DeepEqual(err, tc.expectedError) {
				t.Fatalf("expected error %v, got %v", tc.expectedError, err)
			}

			actualIndexerObjects := map[reflect.Type][]interface{}{}
			for indexerType, indexer := range indexers {
				actualIndexerObjects[indexerType] = indexer.List()
			}

			if !reflect.DeepEqual(actualIndexerObjects, tc.expectedIndexerObjects) {
				t.Errorf("expected indexers' objects differ: %s", cmp.Diff(tc.expectedIndexerObjects, actualIndexerObjects))
			}

		})
	}
}

```

Reply:
To validate if correct key algorithm is chosen validate also if objects can be queried by `<namespace>/<name>` key from returned indexer. Key and type can be a test parameter.

Reply:
@zimnx wouldn't that only test one indexer? Wouldn't it be easier to validate expected indexers against `GetIndexers()` result of each map entry?

Reply:
It depends how you structure test parameters. It could be a collection. I'm fine with different approaches as long as querying via key is validated in unit test.